src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -171,12 +176,12 @@ internal SolutionState Branch(
         SolutionOptionSet? options = null,
         IReadOnlyList<AnalyzerReference>? analyzerReferences = null,
         ImmutableDictionary<string, StructuredAnalyzerConfigOptions>? fallbackAnalyzerOptions = null,
-        ImmutableDictionary<ProjectId, ProjectState>? idToProjectStateMap = null,
+        ImmutableArray<ProjectState>? projectStates = null,
         ProjectDependencyGraph? dependencyGraph = null)
     {
         solutionAttributes ??= SolutionAttributes;
         projectIds ??= ProjectIds;
-        idToProjectStateMap ??= ProjectStates;
+        projectStates = projectStates == null ? ProjectStates : projectStates.Value.Sort(CompareProjectStates);
```

Comment:
So this would mean any time we branch we'll still call this sort method, right? Can we avoid that by only having the cases that add/remove do this sort and the rest of the time can just pass it in?

Reply:
This is changing with the new requirement that callers have to pass this in sorted, but previously this would have only sorted had the caller passed in a new projectStates array.